pkg/util.go

```diff
@@ -129,3 +179,568 @@ func (opt *etcdOptions) setCredentials(sh Shell, appBinding *appcatalog.AppBindi
+		err := errors.New("No pvc found for etcd statefulset")
+		return nil, err
+	}
+
+	restorePvcList := make([]corev1.PersistentVolumeClaim, 0)
+	inv, err := invoker.ExtractRestoreInvokerInfo(opt.kubeClient, opt.stashClient, opt.InvokerKind, opt.InvokerName, opt.namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceRequirement := pvcList[0].Spec.Resources
+
+	if opt.memReq != "" || opt.memLimit != "" {
+		resourceRequirement = corev1.ResourceRequirements{
+			Limits: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse(opt.cpuLimit),
+				corev1.ResourceMemory: resource.MustParse(opt.memLimit),
+			},
+			Requests: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse(opt.cpuReq),
+				corev1.ResourceMemory: resource.MustParse(opt.memReq),
+			},
+		}
+	}
+	if opt.storageClassName == "" {
+		opt.storageClassName = *pvcList[0].Spec.StorageClassName
+	}
+
+	for _, pvcs := range pvcList {
+		pvc := &corev1.PersistentVolumeClaim{
+
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      meta.ValidNameWithPrefix("restore", pvcs.Name),
+				Namespace: opt.namespace,
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				AccessModes:      pvcs.Spec.AccessModes,
+				Resources:        resourceRequirement,
+				StorageClassName: &opt.storageClassName,
+			},
+		}
+
+		createdPVC, _, err := core_util.CreateOrPatchPVC(context.TODO(), opt.kubeClient, pvc.ObjectMeta, func(in *corev1.PersistentVolumeClaim) *corev1.PersistentVolumeClaim {
+			// Set RestoreSession as owner of the PVC so that it get deleted when the respective owner is deleted.
+			core_util.EnsureOwnerReference(&in.ObjectMeta, inv.OwnerRef)
+			in.Spec = pvc.Spec
+			return in
+		}, metav1.PatchOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		restorePvcList = append(restorePvcList, *createdPVC)
+	}
+	return restorePvcList, nil
+}
+
+func (opt *etcdOptions) createRestorePods(pvcList, restorePvcList []corev1.PersistentVolumeClaim) ([]corev1.Pod, []string, error) {
+
+	podList, err := opt.getPodsFromStatefulset()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(podList) == 0 {
+		err := errors.New("No found for etcd statefulset")
+		return nil, nil, err
+	}
+
+	// Getting the invoker info
+	inv, err := invoker.ExtractRestoreInvokerInfo(opt.kubeClient, opt.stashClient, opt.InvokerKind, opt.InvokerName, opt.namespace)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: inv.ObjectMeta.Labels,
+	})
+
+	// Extracting restore job information
+	jobs, err := opt.kubeClient.BatchV1().Jobs(inv.ObjectMeta.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector.String()})
+	klog.Infoln(labels.FormatLabels(inv.ObjectMeta.Labels))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if jobs.Size() == 0 {
+		err = errors.New("No Stash restore job found to get security context for creating restore pods")
+		return nil, nil, err
+	}
+	job := jobs.Items[0]
```

Comment:
Use invoker info for injecting runtimeSettings. We don't need to get the Job here.